src/CommunityToolkit.Maui/Behaviors/MaskedBehavior.shared.cs

```diff
@@ -35,127 +36,135 @@ public class MaskedBehavior : BaseBehavior<InputView>
 	/// </summary>
-	public char UnMaskedCharacter
+	public char UnmaskedCharacter
 	{
-		get => (char)GetValue(UnMaskedCharacterProperty);
-		set => SetValue(UnMaskedCharacterProperty, value);
+		get => (char)GetValue(UnmaskedCharacterProperty);
+		set => SetValue(UnmaskedCharacterProperty, value);
 	}
 
-	static void OnMaskPropertyChanged(BindableObject bindable, object oldValue, object newValue)
-		=> ((MaskedBehavior)bindable).SetPositions();
-
-	static void OnUnMaskedCharacterPropertyChanged(BindableObject bindable, object oldValue, object newValue)
-		=> ((MaskedBehavior)bindable).OnMaskChanged();
-
 	/// <inheritdoc />
-	protected override void OnViewPropertyChanged(object? sender, PropertyChangedEventArgs e)
+	protected override async void OnViewPropertyChanged(InputView sender, PropertyChangedEventArgs e)
 	{
 		base.OnViewPropertyChanged(sender, e);
 
 		if (e.PropertyName == InputView.TextProperty.PropertyName)
 		{
-			OnTextPropertyChanged();
+			await OnTextPropertyChanged();
 		}
 	}
 
-	void OnTextPropertyChanged()
+	static void OnMaskPropertyChanged(BindableObject bindable, object oldValue, object newValue)
 	{
-		if (applyingMask)
-		{
-			return;
-		}
+		var mask = (string?)newValue;
+		var maskedBehavior = (MaskedBehavior)bindable;
 
-		applyingMask = true;
-		ApplyMask(View?.Text);
-		applyingMask = false;
+		maskedBehavior.SetMaskPositions(mask);
 	}
 
-	void SetPositions()
+	static async void OnUnmaskedCharacterPropertyChanged(BindableObject bindable, object oldValue, object newValue)
 	{
-		if (string.IsNullOrEmpty(Mask))
+		var maskedBehavior = (MaskedBehavior)bindable;
+		await maskedBehavior.OnMaskChanged(maskedBehavior.Mask).ConfigureAwait(false);
+	}
+
+	Task OnTextPropertyChanged() => ApplyMask(View?.Text);
+
+	void SetMaskPositions(in string? mask)
+	{
+		if (string.IsNullOrEmpty(mask))
 		{
-			positions = null;
+			maskPositions = null;
 			return;
 		}
 
 		var list = new Dictionary<int, char>();
-		if (Mask != null)
+
+		for (var i = 0; i < mask.Length; i++)
 		{
-			for (var i = 0; i < Mask.Length; i++)
+			if (mask[i] != UnmaskedCharacter)
 			{
-				if (Mask[i] != UnMaskedCharacter)
-				{
-					list.Add(i, Mask[i]);
-				}
+				list.Add(i, mask[i]);
 			}
 		}
 
-		positions = list;
+		maskPositions = list;
 	}
 
-	void OnMaskChanged()
+	async ValueTask OnMaskChanged(string? mask)
 	{
-		if (string.IsNullOrEmpty(Mask))
+		if (string.IsNullOrEmpty(mask))
 		{
-			positions = null;
+			maskPositions = null;
 			return;
 		}
 
-		var originalText = RemoveMaskNullableString(View?.Text);
-		SetPositions();
-		ApplyMask(originalText);
+		var originalText = RemoveMask(View?.Text);
+
+		SetMaskPositions(mask);
+
+		await ApplyMask(originalText);
 	}
 
-	string? RemoveMaskNullableString(string? text)
+	[return: NotNullIfNotNull("text")]
+	string? RemoveMask(string? text)
 	{
-		if (text == null || string.IsNullOrEmpty(text))
+		if (text is null)
```

Comment:
Instead of 2 ifs you can write 1: if (string.IsNullOrEmpty(text)) return text;

Reply:
Nice! 

Yea, I was trying to be too cute by adding `[return: NotNullIfNotNull("text")]`, thinking that the analyzers would need me to check for `null` and return `null` explicitly.

But you're right - this totally works 💯 

```cs
[return: NotNullIfNotNull("text")]
string? RemoveMask(string? text)
{
  if (string.IsNullOrEmpty(text))
  {
	  return text;
  }
// ...
```

Reply:
Updated ✅